ENH: Add support for building with clang-cl

bashtage opened the pull request:
Add support for clang-cl on Windows

NumPy builds and passes with clang-cl.

- [x] Is this a worthwhile effort? Is it useful?  I have found clang-cl to be very helpful in some situations where MSVC won't produce decent code. 
- [x]  Are there a lot of MSVC-specific changes that need to be checked?
- [x]  If this goes forward, should there be a CI build with it?
- [x]  Need to check x86 as well. 
- [x]  Does this need to go to the mailing list?


cc @matthew-brett

bashtage committed:
ENH: Add support for building with clang-cl

Add support for clang-cl on Windows

charris commented:
I think this is useful, I have hopes that clang will eventually give us opensource C and Fortran compilers for windows, especially the latter.

mattip commented on numpy/distutils/ccompiler.py:
```diff
@@ -703,14 +703,17 @@ def CCompiler_cxx_compiler(self):
                                  ('nt', 'intelw'),
-                                 ('nt', 'intelemw'))
+                                 ('nt', 'intelemw'),
+                                 ('nt', 'clangcl'))
```
This could be future-proofed (at least future-improved) by putting the closing paretheses on its own line
```
('nt', 'intelemw'),
('nt', 'clangcl'),
)
```

bashtage replied:
Fixed.

mattip commented:
Can you compile cpython with clangcl on windows? I am a bit surprised we need to add our own ClangCL class to distutils. 

It would indeed be useful to set up a windows CI run under 32-bit and 64-bit for this.

bashtage commented:
> Can you compile cpython with clangcl on windows? I am a bit surprised we need to add our own ClangCL class to distutils.

There is an open ticket with a patch to allow it to build: https://bugs.python.org/issue33351

This said, distutils doesn't support clang-cl any more than distutils supports icc on windows.  It is unfortunate that commands like CC={cl|icl|clang-cl} python setup.py develop don't work Windows like that do on Unix.

AKAICT this is why NumPy would need to directly support clang-cl (if it is a good idea).

mattip commented:
How would this play out for clang on linux, since the new class is windows-specific?

bashtage commented:
clang on unix is trivial since you can use cc=clang python setup.py install.  Windows doesn't honor cc variables.

mattip commented:
What is the magical incantation to get runtests.py or setup.py to use clang on windows?

bashtage commented:
I used the same trick that is used for icc

```
python setup.py config --compiler=clang-cl build_clib --compiler=clang-cl build_ext --compiler=clang-cl install
```

See https://software.intel.com/en-us/articles/numpyscipy-with-intel-mkl

bashtage commented:
It is likely that I'm not doing it correctly.

mattip commented:
`build` already accepts a `compiler` option and passes it on to the sub-commands, so it should be sufficient to run `setup.py build --compiler=clang-cl`. We could add a `--compiler`  option to `runtests.py` that would get passed on to `build`

bashtage commented:
That would probably help.  I think clang-cl might still need a few settings since it isn't 100% identical to cl.

serge-sans-paille commented:
@bashtage : I <3 that PR. If you're interested in extra tests, I can launch the full pythran recompilation based on this, we are in great need for such support, and currently rely on a dirty hack for clang-cl support.

bashtage commented:
I would like to finish this in the next month or so.  Needs a good bit of cleaning I think.

h-vetinari commented:
Any news on this PR? :)

h-vetinari commented:
I hope another ping 4 months later does not constitute bad form. :wink:

Commit:
MAINT: Update clang-cl compiler

Fix compiler for recent changes in NumPy

bashtage commented:
@mattip I've updated to reflect changes and this is working.  I have also improved it.  Any idea how I could take this forward?

bashtage committed:
ENH: Allow compiler to be set in runtests

Add support for --compiler in run tests

bashtage commented:
I think it is there, although ideally it would get a CI run.

```
python runtests.py --compiler=clang-cl
.....................................................................................
.....................................................................................
.....................................................................................
10394 passed, 454 skipped, 108 deselected, 17 xfailed, 1 xpassed in 73.10s (0:01:13)     
````

bashtage committed:
DOC: Improve instructions for building

Clarify methods to build or test

Commit:
MAINT: Make definition match exactly

Make definition and implementation match exactly for clang-cl compat

bashtage committed:
DOC: Add a release note

bashtage commented:
@seberg @mattip @charris Any idea how I can blacklist an intrinsic (or an intrinsic check)?

When building on 32-bit clang-cl has an issue where it generates code using ___mulodi4 which is not available on Windows. This only happens when HAVE___BUILTIN_MUL_OVERFLOW is used. If I manually change the `line `#ifdef HAVE___BUILTIN_MUL_OVERFLOW` to `#if 0` so the fall-back is used, runtests.py completes without any errors.  However, the default creates an error compiling.

Can I just add an #undef in npy_config.py like

```
/* 32-bit clang-cl does not generte correct code for built-in mul overflow
 * https://lists.llvm.org/pipermail/llvm-bugs/2018-March/063437.html
 */
#if defined(_MSC_VER) && defined(__clang__) && !defined(_WIN64)

#undef HAVE___BUILTIN_MUL_OVERFLOW

#endif
```

Commit:
BUG: Block compiling across bitness

Forbid compiling 32 with 64 bit clang-cl and vice versa
Link against built-ins on 32-bit platforms
Improve formatting and add comments

bashtage commented:
Seems like I might be able to link against LLVM\lib\clang\10.0.0\lib\windows\clang_rt.builtins-i386.lib which contains the missing symbol

bashtage commented:
Linking against the builtin lib fixed the issue w/o blacklisting the built-in.

mattip commented:
Could use a CI run. According to the [azure pipelines image](https://github.com/actions/virtual-environments/blob/master/images/win/Windows2019-Readme.md) repo, the windows 2019 image comes with "Microsoft.VisualStudio.Component.VC.Llvm.ClangToolset" but I have no idea if that is what we need.

bashtage committed:
TST: Add Clang build to Azure

bashtage committed:
BUG: Allow clang-cl to sub for msvc

Correct == incorrectly changed to not in

mattip commented on azure-steps-windows.yml:
```diff
@@ -14,7 +14,7 @@ steps:
     # mingw does not like. Instead copy it to a directory and set OPENBLAS,
     # since OPENBLAS will be picked up by the openblas discovery
-    python -m pip  install
+    python -m pip install
```
I think this line was left here by mistake: it should be removed.

bashtage replied:
Thanks.  I couldn't imagine what it was doing.

mattip reviewed (commented):
Is `clang_cl` ever the name of the clang compiler on other platforms? It seems the comparisons `c_compiler.compiler_type in ("msvc", "clang_cl")` might be dangerous if the name is reused.

bashtage commented:
Will switch to draft; getting things to work correctly with blas is proving challenging.

bashtage commented:
Has kind to turned into a bit of a monster. But working now.

pv commented on numpy/distutils/command/build_clib.py:
```diff
@@ -223,7 +223,7 @@ def build_a_library(self, build_info, lib_name, libraries):
             self.mkpath(module_build_dir)
 
-        if compiler.compiler_type == 'msvc':
+        if compiler.compiler_type not in ('msvc', 'clang_cl'):
```
Doesn't this invert the logic?

pv commented on numpy/distutils/fcompiler/gnu.py:
```diff
@@ -215,7 +215,7 @@ def get_libraries(self):
         c_compiler = self.c_compiler
         if sys.platform == 'win32' and c_compiler and \
-                c_compiler.compiler_type == 'msvc':
+                c_compiler.compiler_type not in ("msvc", "clang_cl"):
```
Again `not in` -> `in`?

pv commented on numpy/distutils/clang_cl.py:
```diff
@@ -0,0 +1,161 @@
+    A modified clang-cl compiler compatible with an MSVC-built Python.
+    """
+
+    compiler_type = "clang_cl"
```
Not sure if it matters, but upstream patch seems to set this to `clang-cl`?
https://github.com/python/cpython/pull/18371/files#diff-d962ba8e352edf080a341706a50c13b0R570

bashtage replied:
I have renamed to clang-cl for simplicity.  I think it can be anything, the string is only used in choosing paths.

bashtage committed:
MAINT: Rename clang_cl -> clang-cl

Co-authored-by: Matti Picus <matti.picus@gmail.com>

pv commented on numpy/distutils/ccompiler.py:
```diff
@@ -701,7 +701,7 @@ def CCompiler_cxx_compiler(self):
 compiler_class['intelemw'] = ('intelccompiler', 'IntelEM64TCCompilerW',
                               "Intel C Compiler for 64-bit applications on Windows")
-compiler_class['clang-cl'] = ('clang_cl', 'ClangCL',
+compiler_class['clang-cl'] = ('clang-cl', 'ClangCL',
```
I think this one has to be `clang_cl`.

serge-sans-paille commented:
@bashtage thank you so much for looking into this. If you need some beta-testing, the pythran compiler would be very happy to adopt this clang-cl support, so we can definitively test it when needed :-)

bashtage commented:
@serge-sans-paille It would be helpful to get more experience with this, especially with C++.

mattip commented:
> Is clang_cl ever the name of the clang compiler on other platforms? 

Just to answer my own question: No. "[clang-cl](https://clang.llvm.org/docs/UsersManual.html#clang-cl) is an alternative command-line interface to Clang, designed for compatibility with the Visual C++ compiler, cl.exe."

mattip commented on numpy/distutils/ccompiler.py:
```diff
@@ -701,14 +701,18 @@ def CCompiler_cxx_compiler(self):
 compiler_class['intelemw'] = ('intelccompiler', 'IntelEM64TCCompilerW',
                               "Intel C Compiler for 64-bit applications on Windows")
+compiler_class['clang-cl'] = ('clang_cl', 'ClangCL',
+                              "Clang CL Compiler for 64-bit Windows")
```
Doesn't clang-cl support 32-bit compilation?
```suggestion
                              "Clang CL Compiler for Windows")
```

mattip reviewed (commented):
LGTM. Thanks! The new CI run even does a "full" test. We should put this in soon so other projects can try it out.

bashtage commented:
> Just to answer my own question: No. "[clang-cl](https://clang.llvm.org/docs/UsersManual.html#clang-cl) is an alternative command-line interface to Clang, designed for compatibility with the Visual C++ compiler, cl.exe."

I called it Clang-CL since it is WIndows only. I would not want a Linux user to think they need to use "Clang" rather than just doing a `CC=clang python setup.py ....`

bashtage committed:
BUG: Fix inlining issue

Fix for inlining issue
Try lld-link

bashtage committed:
CLN: Clean up issues creating warnings

bashtage committed:
ENH: Add clang-cl optimization

Add optimization
Revert changes for warnings awaiting #18432
Small clean up of commented code

rgommers commented:
SciPy now uses `clang-cl.exe` in CI for building Pythran extensions, and it seems to work fine on Windows now (I hope I'm not jinxing it by saying that ....). 

This seems close to ready, right @bashtage? The lint failure just prevents seeing if Windows CI build is happy.

> Does this need to go to the mailing list?

Nope, extra compiler support is always accepted.

bashtage commented:
@rgommers Lint is fixed and now green.  It would be good to finally get this in.

charris commented on numpy/core/src/multiarray/array_coercion.c:
```diff
@@ -588,7 +592,11 @@ npy_unlink_coercion_cache(coercion_cache_obj *current)
-NPY_NO_EXPORT NPY_INLINE void
+NPY_NO_EXPORT
+#if (!defined(_MSC_VER) && !defined(__clang__))
+NPY_INLINE
```
Does inlining cause a problem?

bashtage replied:
For some reason clang-cl really didn't like this idiom.

charris replied:
NPY_INLINE is defined in  `npy_common.h` by
```
#if defined(_MSC_VER)
        #define NPY_INLINE __inline
#elif defined(__GNUC__)
    #if defined(__STRICT_ANSI__)
         #define NPY_INLINE __inline__
    #else
         #define NPY_INLINE inline
    #endif
#else
    #define NPY_INLINE
#endif
```
So it should be blank already unless one of  `_MSC_VER` or `__GNUC__`  is defined.

charris replied:
Hmm, I see you have already updated that bit of code.

charris replied:
Looks to me like the `NPY_INLINE` macro should just be removed or the function should be defined in `array_coercion.h` that is included in `ctors.c`. The current code in incorrect C.

bashtage replied:
I have removed these here.  I also submitted a PR to remove these irrespective of this PR

bashtage replied:
clang-cl defines `_MSC_VER`, so `NPY_INLINE` is defined by default.

I removed the inline.  I can't imagine there is a detectable perf different from these even if compilers somehow managed to inline them (they probably ignore it).

bashtage committed:
MAINT: Remove _udiv128 from clang-cl path

_udiv128 is not available in clang-cl
Revert change from llabs to labs

charris commented on numpy/core/src/umath/fast_loop_macros.h:
```diff
@@ -305,7 +305,7 @@ abs_ptrdiff(char *a, char *b)
 #define IS_OUTPUT_BLOCKABLE_UNARY(esizein, esizeout, vsize) \
     ((steps[0] & (esizein-1)) == 0 && \
-     steps[1] == (esizeout) && llabs(steps[0]) < MAX_STEP_SIZE && \
+     steps[1] == (esizeout) && labs(steps[0]) < MAX_STEP_SIZE && \
```
This change seems risky, the function needs to handle `npy_intp`, and long type is untrustworthy for that.

charris replied:
`llabs` is new in C99, so should be available.

charris replied:
Is there a warning about the comparison?

charris replied:
`labs` is used for all the other checks, I think that is incorrect. @seiko2plus Thoughts?

bashtage replied:
There was a warning, which lead to the change.

bashtage replied:
I reverted this and there are no warnings with the latest clang-cl, so I've reverted it.

bashtage commented:
There are two warnings with latest clang-cl

```text
numpy\random\src\distributions\distributions.c(991,13): warning: implicit conversion from 'long long' to 'double' changes value from 9223372036854775807 to 9223372036854775808 [-Wimplicit-const-int-float-conversion]
    if (X > RAND_INT_MAX || X < 1.0) {
          ~ ^~~~~~~~~~~~
numpy\core\include\numpy/random/distributions.h(28,22): note: expanded from macro 'RAND_INT_MAX'
#define RAND_INT_MAX INT64_MAX
                     ^~~~~~~~~
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE\stdint.h(55,26): note: expanded from macro 'INT64_MAX'
#define INT64_MAX        9223372036854775807i64
                         ^~~~~~~~~~~~~~~~~~~~~~
```

and 

```text
numpy\core\src\multiarray\convert_datatype.c(247,9): warning: comparison of nonnull parameter 'dtype' equal to a null pointer is 'false' on first encounter [-Wtautological-pointer-compare]
    if (dtype == NULL) {
        ^~~~~    ~~~~
build\src.win-amd64-3.8\numpy\core\include/numpy\__multiarray_api.h(112,40): note: declared 'nonnull' here
NPY_NO_EXPORT NPY_STEALS_REF_TO_ARG(2) NPY_GCC_NONNULL(2) PyObject * PyArray_CastToType \
                                       ^
numpy\core\include\numpy/npy_common.h(81,43): note: expanded from macro 'NPY_GCC_NONNULL'
#define NPY_GCC_NONNULL(n) __attribute__((nonnull(n)))
                                          ^
1 warning generated.
```

charris commented:
The nonnull attribute is explained here https://www.keil.com/support/man/docs/armcc/armcc_chr1359124976631.htm. AFAICT, the compiler can check if a NULL value is passed, but it isn't clear to me that it is safe to rely on the compiler for that. See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36166. I think the upshot is that it is unsafe to use nonnull unless the programmer is certain that a null cannot be passed in a way that cannot be checked at compile time. These annotations are in numpy/core/code_generators/numpy_api.py and have been present since 2014. My understanding is that with them the compiler is free to remove the NULL check.

bashtage commented:
Unintentional?

bashtage committed:
MAINT: Reduce warnings

Cast to silence warning
Fix type in compiler_type check

mattip commented:
Reopening

seberg commented:
The warning seems incorrect? `dtype` is marked as `nonnull` in the header, that is correct.  But the `Py_SETREF(dtype, ...)` expands to replace `dtype` with something that can be NULL.  Unless marking `nonnull` applies to the body as well?

About the bad inlining, I think I had a reason, but I am sure was wrong, so either way moving to the header or just removing the INLINE is fine by me. If I ever notice it makes a differnece I will just undo it.

seberg commented:
Maybe clang puts the `nonnull` attribute to the definition instead of just constraining that NULL cannot be passed in.   If its annoying, just create a new variable and DECREF the `dtype` passed in afterwards.

charris commented:
> The warning seems incorrect? dtype is marked as nonnull in the header, that is correct. But the Py_SETREF(dtype, ...) expands to replace dtype with something that can be NULL. Unless marking nonnull applies to the body as well?

Might be a compiler bug. Maybe write out the SETREF. Is it necessary to set dtype to NULL if there is an error return?

charris commented:
> Unless marking nonnull applies to the body as well?

Some test cases show that gcc warns on passing NULL, but allows you to set the variable to NULL and do comparisons in the function definition. OTOH, if the compiler is allowed to rewrite the code assuming the parameter is not NULL, it is hard to see how it should be allowed. The problem is that the behavior seems compiler dependent and we don't know what the compilers might actually do.

charris commented:
Note that clang version 11 on Fedora 33 also allows the NULL assignment and comparison without warning.
```
#include <stddef.h>

int* __attribute__((nonnull(1))) f(int *par)
{
    par = NULL;
    if (par == NULL) {
        return NULL;
    }
    return par;
}

int main()
{
    return 0;
}
```

bashtage commented:
I checked clang 10 and 12 on Ubuntu and both produce the same warning about
tautological true.

On Tue, Apr 20, 2021, 20:14 Charles Harris ***@***.***> wrote:

> Note that clang version 11 on Fedora 33 also allows the NULL assignment
> and comparison without warning.
>
> #include <stddef.h>
>
> int* __attribute__((nonnull(1))) f(int *par)
> {
>     par = NULL;
>     if (par == NULL) {
>         return NULL;
>     }
>     return par;
> }
>
> int main()
> {
>     return 0;
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/13816#issuecomment-823534983>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKTSRK5YN6DE5R2D3FENQTTJXHCJANCNFSM4H2R5JYA>
> .
>

charris commented:
I get the warning if I don't do anything with the parameter before the comparison, but if I assign NULL to it there is no warning for the compare. I'm thinking the problem for clang is the `Py_SETREF` macro. Why, I don't know. Perhaps the nonnull analysis is done before macro expansion.

charris commented:
@bashtage Can you try the macro expanded version
```
    PyObject *out;
    PyArray_Descr *tmp = dtype;

    dtype = PyArray_AdaptDescriptorToArray(arr, (PyObject *)dtype);
    Py_DECREF(tmp);
    if (dtype == NULL) {
        return NULL;
    }
```

rgommers commented on numpy/distutils/ccompiler_opt.py:
```diff
@@ -307,7 +312,7 @@ class attribute `conf_features`, also its override
 
         on_x86 = self.cc_on_x86 or self.cc_on_x64
-        is_unix = self.cc_is_gcc or self.cc_is_clang
+        is_unix = self.cc_is_gcc or self.cc_is_clang or self.cc_is_clang_cl
```
`is_unix` will be confusing if it starts including clang-cl on Windows. How about renaming to `is_gcc_or_clang`?

bashtage replied:
I've renamed this.

rgommers commented on numpy/distutils/ccompiler_opt.py:
```diff
@@ -2325,7 +2338,8 @@ def report(self, full=False):
         ))
         ########## baseline ##########
-        if self.cc_noopt:
+        # TODO: Fix this
```
CI was passing - what still needs doing here?

bashtage replied:
Not needed, so removed.

rgommers commented on runtests.py:
```diff
@@ -451,6 +453,8 @@ def build_project(args):
 
     cmd += ["build"]
+    if args.compiler:
+        cmd += [f"--compiler={args.compiler}"]
```
Further down there's a line:
```
compiler = env.get('CC') or cvars.get('CC', '')
```
which then gets used to add some flags. Could you make that use `args.compiler` if defined?

bashtage replied:
Added below.

rgommers commented on numpy/distutils/clang_cl.py:
```diff
@@ -0,0 +1,167 @@
+                install
+
+Develop mode:
+python setup.py build_ext -i --compiler=clang-cl develop
```
Missing a `config` here? `setuptools` shouldn't understand a `--compiler` flag to `setup.py`.

bashtage replied:
Thanks. Fixed now and tested.

rgommers commented on numpy/distutils/clang_cl.py:
```diff
@@ -0,0 +1,167 @@
+        # Forbid compiling 32 on 64-bit LLVM and vice versa
+        target = "i686" if platform_bits == 32 else "x86_64"
+        if target not in clang_full_ver:
+            raise RuntimeError(
```
This makes sense. Would be good to add to the comment that `platform_bits` represents the Python interpreter, not the OS.

bashtage replied:
Comment expanded.

rgommers commented on numpy/distutils/clang_cl.py:
```diff
@@ -0,0 +1,167 @@
+        self.include_dirs.insert(0, clang_incl)
+
+        # Need to block MS includes except those from Visual Studio 14.0
+        # since they include intrinsics that break clang
```
This sounds like something that'll come back to bite us. Could you expand the comment a bit? 14.0 is Visual Studio 2015. Is anything >= 15.0 (or >= 2017) to be avoided?

rgommers replied:
If there's a public issue somewhere, that'd be helpful.

bashtage replied:
I've linked to a SO question.  I believe ucrt stuff is always in 14.0 irrespective of the actual compiler used.  Will need to block newer compilers as well.

rgommers commented on numpy/distutils/clang_cl.py:
```diff
@@ -0,0 +1,167 @@
+        # Need to block MS includes except those from Visual Studio 14.0
+        # since they include intrinsics that break clang
+        def retain_include(path):
+            blocked = ("Windows Kits", "MSVC", "2019")
```
Is it guaranteed that the path for Visual Studio 14.0 doesn't include "MSVC"?

bashtage replied:
This is definitely the most challenging part of this.  I've linked to a SO question that had the relevant details:

https://stackoverflow.com/questions/36086328

rgommers replied:
Thanks, very useful discussion.

rgommers reviewed (commented):
This looks really good. I just have some small comments on the distutils parts.

Commit:
CLN: Improve comments and resolve small issues

rgommers commented on azure-pipelines.yml:
```diff
@@ -274,6 +274,14 @@ stages:
+            BITS: 64
+            NPY_USE_BLAS_ILP64: '1'
+            OPENBLAS_SUFFIX: '64_'
+            CLANG_CL: 1
```
It is surprisingly slow to build - ~3.5x slower than MSVC in CI. Is that expected?

bashtage replied:
Locally it is also slower than clang on unix. OTOH it probably does a better job optimizing some numeric code than msvc.

bashtage replied:
It was building twice (using build and build_ext)-- now 4:40 vs 2:59, so not that much slower.

rgommers replied:
Ah, that's much better, thanks.

seiko2plus commented on numpy/distutils/ccompiler_opt.py:
```diff
@@ -307,35 +312,40 @@ class attribute `conf_features`, also its override
 
         on_x86 = self.cc_on_x86 or self.cc_on_x64
-        is_unix = self.cc_is_gcc or self.cc_is_clang
+        is_gcc_or_clang = (
```
if clang-cl is a Unix-like compiler then why not just treat it as a clang?

bashtage replied:
> if clang-cl is a Unix-like compiler(same compiler options) then why not just treat it as a clang?

It mostly uses cl flags and switches for older choices (e.g., optimization levels), even if it supports a lot of the same compiler-specific syntax as clang/gcc.

seiko2plus replied:
what I can see in your patch that only OPT level `/O2` is required, which can be achieved via:
```Diff
         clang = dict(
             native = '-march=native',
-            opt = "-O3",
+            opt = f"{'/O2 ' if os.name == 'nt' else ''}-O3",
             werror = '-Werror'
         ),
         icc = dict(
```
note any unsupported flags will be ignored during the class initialization, see
https://github.com/numpy/numpy/blob/2a2dc51aacab13eb2ebe3ca97cf1bc674ba7f39f/numpy/distutils/ccompiler_opt.py#L982-L990

bashtage replied:
There are quite a few other specializations needed where clang-cl differs from using CC=clang on *nix, such as customization of the default include paths and linker. Calling it clang-cl clarifies that this is specialized to the specific version of clang that is designed to mirror MSVC and has ABI compatibility with MSVC.

seiko2plus replied:
I agree, having a special definition for `clang-cl` might it be necessary in the future. Thanks.

bashtage commented:
> Can you try the macro expanded version

@charris  If I manually replace the macro it compiles without a warning.  Seems to guarantee that this is a compiler bug that probably only generates a false warning

bashtage committed:
MAINT: Remove hasattr check

Remove check which isn't needed when calling setup with arguments
in the correct order

bashtage commented on numpy/distutils/ccompiler_opt.py:
```diff
@@ -307,35 +312,40 @@ class attribute `conf_features`, also its override
-            AVX512_CNL = dict(flags="-mavx512ifma -mavx512vbmi"),
-            AVX512_ICL = dict(
-                flags="-mavx512vbmi2 -mavx512bitalg -mavx512vpopcntdq"
+        if on_x86 and is_gcc_or_clang:
```
These changes are brought about by the lint check since you have to correct both lines that are changed and +/-1 line too. This spilled over to the entire dictionary.

seiko2plus replied:
Thank you for the clarification

bashtage committed:
BUG: Remove inline decorator from shared function

bashtage commented:
@seberg I needed to remove the NPY_INLINE here as I did previously from `array_coercion.c`.  While I'm not 100% on this, I think no-export inline functions like this need to be in the header so that they are available to other translation units.

bashtage committed:
Fix distributor init

bashtage committed:
Continue to work on issues in Clang-cl

bashtage committed:
BUG: Use workaround on clang-cl

Bug continues on clang-cl

bashtage commented:
@seberg @mattip Do you know where left shift is implemented that would be used with a vector?  

I have fixed a few issues that have arisen due to vectorized functions introduced, but still have two failures related to "RuntimeWarning: invalid value encountered in left_shift"

```python
import numpy as np

dt = np.dtype('int32')
op = np.left_shift
shift_arr  = np.array([32] * 32, dtype=dt)
val_arr    = np.array([5] * 32, dtype=dt)
op(val_arr, shift_arr)
```

I have a feeling that this is related to the `CLANG_RECIP_WORKAROUND` flag that is also needed in the PR, but I thought it would be good to see if I could find the implementation to write a pure C implementation that produces the same.

If I run this code I only see it when m >= 2, which suggests to me there is an SSE/SSE2 implementation somewhere.

```
for m in range(6):
     shift_arr  = np.array([32]*2**m, dtype=dt)
     val_arr = np.array([5]*2**m, dtype=dt)
     print(m)
     op(val_arr, shift_arr)
```

InessaPawson commented:
Do we wish to continue working on this PR, given numpy.distutils has been officially deprecated?

bashtage commented:
The PR is eol. I think the good parts all got merged in other PRs.

mattip commented:
Let's close this
